src/de.rs

```diff
@@ -398,6 +400,50 @@ impl<S: Read> Decoder<S> {
         Ok(i)
     }
 
+    pub(crate) async fn peek(&mut self) -> Result<Option<u8>, Error> {
+        while self.buffer.is_empty() && !self.source.is_terminated() {
+            self.buffer().await?;
+        }
+
+        if self.buffer.is_empty() {
+            Ok(None)
+        } else {
+            Ok(Some(self.buffer[0]))
+        }
+    }
+
+    async fn peek_or_null(&mut self) -> Result<u8, Error> {
+        Ok(self.peek().await?.unwrap_or(b'\x00'))
```

Comment:
I think it would be better to check if `peek` returns `Some(...)` since AFAIK the null byte has no significance in the JSON spec

Reply:
these methods i copied over from serde - i think you're right though.  i've removed both of these methods, and refactored the usages to just use `peek()` or `next_char()`